### PR TITLE
Logging macros

### PR DESCRIPTION
**Pull request by kshxtij**
Add macros for logging something only once or n times to reduce spam in logs

### CODE REVIEW

File: lib/core/logger.hpp [hunk 1]
```diff
@@ -6,6 +6,17 @@
...
 #include <optional>
 
+#define logNTimes(logger, n, level, ...)                                                           \
+  do {                                                                                             \
+    static int count = 0;                                                                          \
+    if (count < n) {                                                                               \
+      logger.log(level, __VA_ARGS__);                                                              \
+      count++;                                                                                     \
+    }                                                                                              \
+  } while (0)
```
[the remark this answers is not in the document]
Comment (kshxtij) [reply lib/core/logger.hpp:16]: We don't want it to run n times, that exists to group multiple lines of code into a single macro

[the remark this answers is not in the document]
Comment (SnickeyX) [reply lib/core/logger.hpp:16]: `static int count` remaining `static` throughout breaks me but yes

**Review by SnickeyX — APPROVED** [review]

File: lib/core/logger.hpp [hunk 2]
```diff
@@ -6,6 +6,17 @@
...
+
+#define logOnce(logger, level, ...) logNTimes(logger, 1, level, __VA_ARGS__)
```
Comment (mifrandir) [inline lib/core/logger.hpp:18]: what's this?

File: lib/core/logger.hpp [hunk 3]
```diff
@@ -6,6 +6,17 @@
 #include <iostream>
 #include <optional>
 
+#define logNTimes(logger, n, level, ...)                                                           \
+  do {                                                                                             \
+    static int count = 0;                                                                          \
+    if (count < n) {                                                                               \
+      logger.log(level, __VA_ARGS__);                                                              \
+      count++;                                                                                     \
```
Comment (mifrandir) [inline lib/core/logger.hpp:14]: prefix

File: lib/core/logger.hpp [hunk 4]
```diff
@@ -6,6 +6,17 @@
 #include <iostream>
 #include <optional>
 
+#define logNTimes(logger, n, level, ...)                                                           \
+  do {                                                                                             \
+    static int count = 0;                                                                          \
```
Comment (mifrandir) [inline lib/core/logger.hpp:11]: `std::size_t`

**Review by mifrandir — CHANGES_REQUESTED** [review]

**Comment by kshxtij** [comment]
@mifrandir could i get a review whenever you have time?

**Review by mifrandir — APPROVED** [review]
another cool thing would be "log every nth time"

File: lib/core/logger.hpp [hunk 5]
```diff
@@ -15,6 +15,13 @@
     }                                                                                              \
   } while (0)
 
+#define logEveryNth(logger, n, level, ...)                                                         \
+  do {                                                                                             \
+    static std::size_t count = 0;                                                                  \
+    if (count % n == 0) { logger.log(level, __VA_ARGS__); }                                        \
```
Comment (mifrandir) [inline lib/core/logger.hpp:21]: check `count == n` and reset the counter to avoid overflows

File: test/core/logger.cpp [hunk 6]
```diff
@@ -58,4 +58,34 @@ TEST(Logger, VaryingTimes)
...
+
+TEST(Logger, logEveryNth)
+{
+  utils::ManualTime manual_time;
+  setenv("TZ", "Europe/London", 1);
+  tzset();
+  testing::internal::CaptureStdout();
+  core::Logger logger("test", core::LogLevel::kDebug, manual_time);
+  for (int i = 0; i < 15; i++) {
+    logEveryNth(logger, 5, core::LogLevel::kDebug, "test");
```
Comment (mifrandir) [inline test/core/logger.cpp:84]: advance manual time to make sure the right logs are printed?

**Review by mifrandir — CHANGES_REQUESTED** [review]

File: test/core/logger.cpp [hunk 7]
```diff
@@ -58,4 +58,34 @@ TEST(Logger, VaryingTimes)
...
+}
+
+TEST(Logger, logEveryNth)
+{
+  utils::ManualTime manual_time;
+  setenv("TZ", "Europe/London", 1);
+  tzset();
+  testing::internal::CaptureStdout();
+  core::Logger logger("test", core::LogLevel::kDebug, manual_time);
+  for (int i = 0; i < 15; i++) {
```
Comment (maxguy2001) [inline test/core/logger.cpp:83]: size_t and pre-incrementing

File: test/core/logger.cpp [hunk 8]
```diff
@@ -58,4 +58,34 @@ TEST(Logger, VaryingTimes)
...
 }
 
+TEST(Logger, logNTimes)
+{
+  utils::ManualTime manual_time;
+  setenv("TZ", "Europe/London", 1);
+  tzset();
+  testing::internal::CaptureStdout();
+  core::Logger logger("test", core::LogLevel::kDebug, manual_time);
+  for (int i = 0; i < 5; i++) {
```
Comment (maxguy2001) [inline test/core/logger.cpp:68]: size_t and pre-incrementing

**Review by maxguy2001 — COMMENTED** [review]
Quick mac compliation fix and arbitrary style-guide changes but otherwise all good.

**Review by mifrandir — APPROVED** [review]
modulo max's changes

**Comment by kshxtij** [comment]
This builds locally, i'll track down what's causing CI to fail and merge this later today.

**Comment by maxguy2001** [comment]
> This builds locally, i'll track down what's causing CI to fail and merge this later today.

It's the macos security thing. Merging master into this branch should fix it (directly fixing problem will give you merge conflict).

**Pull request merged**